### PR TITLE
Add io.gitlab.o20.word

### PR DESCRIPTION
**Pull request by abstractdevelop**
We would like to distribute the first O20 app on Flatpak. It is the Office365 clone for Linux, and is already available on the Snap Store (https://snapcraft.io/office20) and has reached a lot of users there. This is our first flatpak, and may not be perfect...

### CODE REVIEW

**Comment by LeandroStanger** [comment]
bot, build,org.o20.word

File: org.o20.word.json [hunk 1]
```diff
@@ -0,0 +1,22 @@
+{
+    "id": "org.o20.word",
+    "branch": "master",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.12",
```
Comment (LeandroStanger) [inline org.o20.word.json]: Update runtime to 5.13

**Comment by LeandroStanger** [comment]
bot, build org.o20.word

**Comment by LeandroStanger** [comment]
bot, build org.o20.word

File: org.o20.word.json [hunk 2]
```diff
@@ -0,0 +1,35 @@
+{
+    "id": "org.o20.word",
+    "branch": "master",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.13",
+    "sdk": "org.kde.Sdk",
+    "command": "word64.run",
+    "tags": ["beta"],
+    "desktop-file-name-suffix": " Beta",
+    "finish-args": ["--share=ipc", "--socket=x11", "--socket=wayland", "--filesystem=host" ],
```
Comment (barthalion) [inline org.o20.word.json]: What does it need host access for? Does it not use portals for opening/saving files?

File: org.o20.word.json [hunk 3]
```diff
@@ -0,0 +1,35 @@
+{
+    "id": "org.o20.word",
+    "branch": "master",
```
Comment (barthalion) [inline org.o20.word.json]: This should be deleted, Flathub uses `stable` as the default branch.

File: org.o20.word.json [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.13",
+    "sdk": "org.kde.Sdk",
+    "command": "word64.run",
+    "tags": ["beta"],
```
Comment (barthalion) [inline org.o20.word.json]: Flathub in general targets stable builds, or at least ones that are considered good enough for daily use by application developers. I'd delete mentions of "beta", unless you would prefer to merge your app only to `beta` branch for now.

File: org.o20.word.json [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+            "builddir": true,
+            "sources": [
+                {
+                    "type": "file",
+                    "path": "org.o20.word.appdata.xml"
+                },
+                {
+                     "type": "git",
+		     "url": "https://gitlab.com/abstractsoftware/o20/o20coreapps.git",
+		     "branch": "master"
```
Comment (barthalion) [inline org.o20.word.json]: It should point to a tag or commit to keep build reproducible.

File: org.o20.word.appdata.xml [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<component type="desktop">
+    <id>org.o20.word</id>
+    <launchable type="desktop-id">org.o20.word.desktop</launchable>
+    <metadata_license>CC0-1.0</metadata_license>
+    <project_license>GPL-2.0+</project_license>
+
+    <name>O20.Word</name>
+
+    <categories>
```
Comment (barthalion) [inline org.o20.word.appdata.xml]: Categories are taken from desktop launcher.

File: org.o20.word.appdata.xml [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+
+    <developer_name>the Abstract Developers</developer_name>
+
+    <releases>
+      <release date="2020-01-22" version="20.1.3"/>
+    </releases>
+
+    <url type="homepage">https://abstractsoftware.gitlab.io</url>
+    <url type="bugtracker">https://gitlab.com/abstractsoftware/o20/o20coreapps/issues</url>
+</component>
```
Comment (barthalion) [inline org.o20.word.appdata.xml]: Add [OARS ratings](https://hughsie.github.io/oars/).

File: CONTRIBUTING.md [hunk 8]
```diff
@@ -1,15 +0,0 @@
-### Submitting a new package
```
Comment (barthalion) [inline CONTRIBUTING.md]: If you rebase your branch on the `new-pr` branch, these files will be gone along with unrelated commits in the history.

File: org.o20.word.appdata.xml [hunk 9]
```diff
@@ -0,0 +1,28 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (barthalion) [inline org.o20.word.appdata.xml]: Ideally this file should be maintained in your git repository and installed by cmake.

**Comment by nedrichards** [comment]
bot, build org.o20.word

**Comment by nedrichards** [comment]
Appdata also needs screenshots.

**Comment by abstractdevelop** [comment]
Thanks @barthalion and @nedrichards, I updated my JSON to do most of the things you said, and moved the AppData to the global repository.

**Comment by nedrichards** [comment]
bot, build org.o20.word

@abstractdevelop you can use that command as well if you want to generate a test build when you push updates.

**Comment by abstractdevelop** [comment]
bot, build org.o20.word

File: org.o20.word.json [hunk 10]
```diff
@@ -0,0 +1,25 @@
+{
+    "id": "org.o20.word",
```
Comment (nedrichards) [inline org.o20.word.json]: so this certainly isn't going to be the "right" app-id unless you own the o20.org website. see https://github.com/flathub/flathub/wiki/App-Requirements#Application-ID for more.

File: org.o20.word.json [hunk 11]
```diff
@@ -0,0 +1,25 @@
...
+    "finish-args": ["--share=ipc", "--socket=x11", "--socket=wayland", "--filesystem=host" ],
+    "separate-locales": false,
+    "modules": [
+        {
+            "name": "o20coreapps",
+            "buildsystem": "cmake-ninja",
+            "builddir": true,
+            "sources": [
+                {
+                     "type": "git",
```
Comment (nedrichards) [inline org.o20.word.json]: can you fix the indentig here at the same time you pin the release to a specific version.

**Comment by nedrichards** [comment]
So, you certainly need to add the `--device=dri` permission go Qt can access the GPU also when opening the test I get an immediate crash:

> double free or corruption (out)

File: org.o20.word.json (hunk 10 above)
Comment (abstractdevelop) [reply org.o20.word.json]: Would io.gitlab.o20.word be better?

**Comment by abstractdevelop** [comment]
OK, did all those things.

**Comment by abstractdevelop** [comment]
bot, build org.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
It's still crashing even after I added the `--device=dri` permission.

**Comment by barthalion** [comment]
You can install debug symbols with `flatpak install --user word-origin io.gitlab.o20.word.Debug` and then `flatpak run --devel --command=bash io.gitlab.o20.word` to poke around. GDB gave me this, but not sure how helpful it is:

```
0x00007ffff5ca4605 in raise () from /usr/lib/x86_64-linux-gnu/libc.so.6
(gdb) bt
#0  0x00007ffff5ca4605 in raise () from /usr/lib/x86_64-linux-gnu/libc.so.6
#1  0x00007ffff5c8d899 in abort () from /usr/lib/x86_64-linux-gnu/libc.so.6
#2  0x00007ffff5ce85bf in ?? () from /usr/lib/x86_64-linux-gnu/libc.so.6
#3  0x00007ffff5cefb7c in ?? () from /usr/lib/x86_64-linux-gnu/libc.so.6
#4  0x00007ffff5cf19e0 in ?? () from /usr/lib/x86_64-linux-gnu/libc.so.6
#5  0x000055555556c8c7 in QTypedArrayData<unsigned short>::deallocate (data=<optimized out>)
    at /usr/include/QtCore/qarraydata.h:236
#6  QString::~QString (this=0x7fffffffdee8, __in_chrg=<optimized out>)
    at /usr/include/QtCore/qstring.h:1133
#7  CreateBreezePalette () at ../dev/o20.word/Main.cpp:41
#8  0x000055555556bacb in main (argc=<optimized out>, argv=<optimized out>)
    at ../dev/o20.word/Main.cpp:110
```

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by LeandroStanger** [comment]
bot, build io.gitlab.o20.word

**Comment by bilelmoussaoui** [comment]
The application crashes whenever I try to type something, here's the bt I got
 ```
Thread 1 "word64.run" received signal SIGSEGV, Segmentation fault.
0x000055555556c672 in QString::~QString() ()
(gdb) bt
#0  0x000055555556c672 in QString::~QString() ()
#1  0x000055555556a4bc in ?? ()
#2  0x00007ffff774fcb1 in QApplicationPrivate::notify_helper(QObject*, QEvent*)
    () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#3  0x00007ffff775995e in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#4  0x00007ffff62f1ad2 in QCoreApplication::notifyInternal2(QObject*, QEvent*)
    () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#5  0x00007ffff77b0c8e in ?? ()
   from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#6  0x00007ffff774fcc2 in QApplicationPrivate::notify_helper(QObject*, QEvent*)
    () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#7  0x00007ffff7759240 in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#8  0x00007ffff62f1ad2 in QCoreApplication::notifyInternal2(QObject*, QEvent*)
    () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#9  0x00007ffff6761df3 in QGuiApplicationPrivate::processKeyEvent(QWindowSystemInterfacePrivate::KeyEvent*) () from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#10 0x00007ffff6767241 in QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQt5Gui.so.5
#11 0x00007ffff06cbb80 in ?? ()
   from /usr/lib/plugins/platforminputcontexts/libibusplatforminputcontextplugin--Type <RET> for more, q to quit, c to continue without paging--
.so
#12 0x00007ffff631cee8 in QMetaObject::activate(QObject*, int, int, void**) ()
   from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#13 0x00007ffff2fc1a9f in QDBusPendingCallWatcher::finished(QDBusPendingCallWatcher*) () from /usr/lib/x86_64-linux-gnu/libQt5DBus.so.5
#14 0x00007ffff2fc1ba0 in ?? () from /usr/lib/x86_64-linux-gnu/libQt5DBus.so.5
#15 0x00007ffff631d79a in QObject::event(QEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#16 0x00007ffff774fcc2 in QApplicationPrivate::notify_helper(QObject*, QEvent*)
    () from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#17 0x00007ffff7759240 in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQt5Widgets.so.5
#18 0x00007ffff62f1ad2 in QCoreApplication::notifyInternal2(QObject*, QEvent*)
    () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#19 0x00007ffff62f4768 in QCoreApplicationPrivate::sendPostedEvents(QObject*, int, QThreadData*) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#20 0x00007ffff6348c03 in ?? () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#21 0x00007ffff31ee033 in g_main_context_dispatch ()
   from /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#22 0x00007ffff31ee2c0 in ?? () from /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#23 0x00007ffff31ee363 in g_main_context_iteration ()
   from /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0
#24 0x00007ffff6348241 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop:--Type <RET> for more, q to quit, c to continue without paging--
:ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#25 0x00007ffff62f063b in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#26 0x00007ffff62f82a2 in QCoreApplication::exec() ()
   from /usr/lib/x86_64-linux-gnu/libQt5Core.so.5
#27 0x000055555556be5a in ?? ()
#28 0x00007ffff5c8f173 in __libc_start_main ()
   from /usr/lib/x86_64-linux-gnu/libc.so.6
#29 0x000055555556c26e in ?? ()
```

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
OK, I fixed that too. But why it is crashing in the global flatpak when it never crashed in the one I built myself?!?

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by nedrichards** [comment]
bot, build io.gitlab.o20.word

**Comment by nedrichards** [comment]
This version builds and runs fine for me. Needs some work on HiDPI but that's outside the scope of this review ;-)

I see
```
org.kde.solid.udisks2: Failed enumerating UDisks2 objects: "org.freedesktop.DBus.Error.Disconnected" 
 "Not connected to D-Bus server"
```
Which may be because I'm not running it in a KDE environment, but if you need to whitelist any dbus address' you use then you need to add that into the manifest. A useful tool for that can be running the flatpak with `--log-session-bus` it's pretty verbose but lets you know everything.

**Comment by barthalion** [comment]
Could be also because of old flatpak-builder. Since few releases it pulls build flags from `/etc/flatpak-builder/defaults.json` file in the Sdk, which at the moment looks like this:

```
{
  'libdir': '/usr/lib/x86_64-linux-gnu',
  'cflags': '-O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection',
  'cxxflags': '-O2 -g -pipe -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -fexceptions -fstack-protector-strong -grecord-gcc-switches -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection',
  'cppflags': '',
  'ldflags': '-L/app/lib -Wl,-z,relro,-z,now -Wl,--as-needed'
}
```

Possibly some of these flags exposed bugs.

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

File: io.gitlab.o20.word.json [hunk 12]
```diff
@@ -0,0 +1,28 @@
+{
+    "id": "io.gitlab.o20.word",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.13",
+    "sdk": "org.kde.Sdk",
+    "command": "word64.run",
+    "tags": ["standalone"],
+    "desktop-file-name-suffix": " (Standalone)",
```
Comment (bilelmoussaoui) [inline io.gitlab.o20.word.json]:
```suggestion
```

File: io.gitlab.o20.word.json [hunk 13]
```diff
@@ -0,0 +1,28 @@
+{
+    "id": "io.gitlab.o20.word",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.13",
+    "sdk": "org.kde.Sdk",
+    "command": "word64.run",
+    "tags": ["standalone"],
```
Comment (bilelmoussaoui) [inline io.gitlab.o20.word.json]: Not used at all here on Flathub, just remove it

File: io.gitlab.o20.word.json [hunk 14]
```diff
@@ -0,0 +1,28 @@
...
+    "modules": [
+        {
+            "name": "o20coreapps",
+            "buildsystem": "cmake-ninja",
+            "builddir": true,
+            "sources": [
+                {
+                     "type": "git",
+                     "url": "https://gitlab.com/abstractsoftware/o20/o20coreapps.git",
+                     "branch": "master",
```
Comment (bilelmoussaoui) [inline io.gitlab.o20.word.json]: Flathub only hosts stable applications, this should be a tagged release or a tarball.

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

File: io.gitlab.o20.word.json [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+    "modules": [
+        {
+            "name": "o20coreapps",
+            "buildsystem": "cmake-ninja",
+            "builddir": true,
+            "sources": [
+                {
+                     "type": "git",
+                     "url": "https://gitlab.com/abstractsoftware/o20/o20coreapps.git",
+		     "tag": "20.0.5"
```
Comment (bilelmoussaoui) [inline io.gitlab.o20.word.json]: Nitpick: indentation issue here. Also, a tag should go with a commit.

File: io.gitlab.o20.word.json [hunk 16]
```diff
@@ -0,0 +1,25 @@
+{
+    "id": "io.gitlab.o20.word",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.13",
+    "sdk": "org.kde.Sdk",
+    "command": "word64.run",
+    "finish-args": ["--share=ipc", "--socket=x11", "--socket=wayland", "--device=dri", "--filesystem=host"],
```
Comment (bilelmoussaoui) [inline io.gitlab.o20.word.json]: Any reason for requiring host access? This went unanswered in the first review.

Comment (bilelmoussaoui) [reply io.gitlab.o20.word.json]: Also, mind putting each permission per line? :)

Comment (abstractdevelop) [reply io.gitlab.o20.word.json]: No, I don't think I need host access. I just though it was why something was happening when I was fixing the seg faults, but now I see I was wrong.

**Comment by bilelmoussaoui** [comment]
bot, build io.gitlab.o20.word

**Comment by bilelmoussaoui** [comment]
After giving the app a try, it assumes that it can writes to '$HOME/O20`
Screenshot from 2020-01-30 21-38-32
I would say, you either need to give it access to `$HOME/O20` or disable the auto save functionality somehow as it won't work.

I have tried clicking on save/save us PDF; but nothing shows up and nothing in the logs.

**Comment by abstractdevelop** [comment]
Do you mean that it doesn't work when saving to $HOME/O20? It works fine on my systems (KDE Neon & Ubuntu Budgie 19.10), and is able to save to the default location, and save as/save as PDF.

**Comment by bilelmoussaoui** [comment]
Well, it's not possible for an app without the permission to access HOME to write directly to it .

**Comment by abstractdevelop** [comment]
OK. I'll disable the AutoSave. But is there a way to give the app permissions to write to HOME?

**Comment by bilelmoussaoui** [comment]
Yes, adding `--filesystem=home/O20` to `finish-args` should do it.

**Comment by abstractdevelop** [comment]
Do you think it would be better to disable the AutoSave, or should I just update the `finish-args`?

**Comment by bilelmoussaoui** [comment]
Just update the finish-args, except, I wouldn't save them directly in `home/O20` but rather in `xdg-documents/O20`

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by abstractdevelop** [comment]
Unfortunately, since O20 create the xdg-documents/O20 directory _if it doesn't exist_, then it can't create that folder unless it has access to xdg-documents, not only xdg-documents/O20. So then should I make it so it saves only to the Documents folder, or give it access to the Documents folder?

**Comment by abstractdevelop** [comment]
Unfortunately, since O20 create the xdg-documents/O20 directory if it doesn't exist, then it can't create that folder unless it has access to xdg-documents, not only xdg-documents/O20. So then should I make it so it saves only to the Documents folder, or give it access to the Documents folder but keep everything else the same?

File: io.gitlab.o20.word.json [hunk 17]
```diff
@@ -0,0 +1,26 @@
+{
+    "id": "io.gitlab.o20.word",
+    "runtime": "org.kde.Platform",
+    "runtime-version": "5.13",
+    "sdk": "org.kde.Sdk",
+    "command": "word64.run",
+    "finish-args": ["--share=ipc", "--socket=x11", "--socket=wayland", "--device=dri", "--filesystem=xdg-documents/O20"],
```
Comment (TingPing) [inline io.gitlab.o20.word.json]: > "--filesystem=xdg-documents/O20"

You'll want `--filesystem=xdg-documents/O20:create` as the directory generally won't exist.

**Comment by abstractdevelop** [comment]
bot, build io.gitlab.o20.word

**Comment by barthalion** [comment]
bot, build io.gitlab.o20.word
